(CFACT-227) Move locale and catch to leatherman

branan opened the pull request:
These are a bit lighter-weight than the logging library, so are a
quicker start for getting it all put together.

This can be tested against https://github.com/branan/cfacter/tree/imp/master/CFACT-227-leatherman

MikaelSmith commented on CMakeLists.txt:
```diff
@@ -0,0 +1,127 @@
+# * automatic test integration with the top-level test executable
+# * variables passed into each subdir for common naming conventions
+# * options to enable/disable individual libraries
+macro(add_leatherman_dir dir)
```
Can we pull everything about add_leatherman_dir into a separate CMake module?

branan replied:
definitely

MikaelSmith commented on catch/CMakeLists.txt:
```diff
@@ -0,0 +1,2 @@
+# Just point the user at our vendored catch. Nothing else to do here
```
What's the value in this redirect?

branan replied:
Avoiding a special-case around catch in the top-level CMake file

MikaelSmith replied:
Alright, I wonder about having the actual source over in vendor/catch then. We could point out that it's vendored in some other way.

MikaelSmith replied:
I see why this makes sense (we may eventually have vendored libraries that we don't plan to export), and the vendor dir pattern is a useful identifier. So this is fine.

MikaelSmith commented on CMakeLists.txt:
```diff
@@ -0,0 +1,127 @@
+endmacro()
+
+# This macro gives us a pile of common variable names and behaviors
+# for leatherman libraries that we can work with
```
The comments should point out the pattern used for enabling libraries.

MikaelSmith commented on CMakeLists.txt:
```diff
@@ -0,0 +1,127 @@
+
+if(LEATHERMAN_ENABLE_TESTING)
+    enable_testing()
+    add_subdirectory(tests)
```
I kind of like having tests in each individual library, so it's obvious which libraries are covered.

branan replied:
That's the intended pattern. This is just for the test executable. I'll rename the directory and clarify that with a comment down here.

MikaelSmith replied:
Oh I see, each library adds a test lib.

MikaelSmith commented:
I don't know if specifying install files makes more sense in this project or in the projects that use it. I lean towards leaving it out here until we see a reason to change it.

MikaelSmith commented:
We should add a README on how to use and include libraries.

branan commented:
@MikaelSmith it's probably worth having a way to build this standalone (as a pile of static libraries and headers), but I don't see it as needed for the initial conversion.

README is in-progress.

branan commented:
@MikaelSmith big update!

* Added README with usage instructions as well as a basic primer on adding a library
* Moved all the macros into their own `.cmake` file
* Cleaned up macro documentation

MikaelSmith commented on README.md:
```diff
@@ -0,0 +1,84 @@
+In addition, each enabled library sets a number of library-specific
+variables:
+
+* `LEATHERMAN_<LIBRARY>_INCLUDE` The include directoryor directories
```
Need a space in `directoryor`.

MikaelSmith commented:
This looks great. :+1:

MikaelSmith commented:
Is this still WIP? Looks ready to merge to me.

kylog commented on cmake/leatherman.cmake:
```diff
@@ -0,0 +1,163 @@
+# Helper Macros
+####
+
+# Usage: debug("Somethign cool is happening")
```
'Something'

kylog commented:
:+1: 

Great work!

kylog commented:
@parisiale @ploubser @richardc thoughts?

ploubser commented:
Sorry, commented on #2 This looks really cool. :+1:

richardc commented:
Seems OK, looking forward to the boost-style `Include(leatherman logging catch)` style of declaring which modules are wanted.

parisiale commented:
Leatherman looks great.
If we agree, I can make a PR for adding the thread ID and source line to the logging format; here's how it would look like: http://git.io/AFV6.

branan commented:
@MikaelSmith @kylog Typos addressed. I think this is ready to be merged now

dvhwgumby commented:
Sorry to add to a four year old open issue, but I cannot use the Leatherman::locale component in CMake.  `find_package(Leatherman REQUIRED locale util)` works, and I can build my code by putting `${LEATHERMAN_LIBRARIES}` in the appropriate `target_link_libraries` entries.

But if I try to do things the "correct" way (by putting `Leatherman::locale` in `target_link_libraries`, CMake complains that the target was not found.  My CMake-fu is inadequate to debug this.

I pulled on this thread while beginning to work on a pull request for cpp-hocon issue 103 (not sure how to link references across github projects).

MikaelSmith commented:
I guess I'm not sure why you say putting `Leatherman::local` in `target_link_libraries` is the correct way? With CMake, you can add something like `message("Libs: ${LEATHERMAN_LIBRARIES}` after `include(leatherman)` to see how the variable's initialized. For cpp-hocon it appears to be `leatherman_util;leatherman_locale;/usr/local/lib/libboost_locale-mt.dylib;/usr/local/lib/libboost_system-mt.dylib;/usr/local/lib/libboost_date_time-mt.dylib;/usr/local/lib/libboost_chrono-mt.dylib;leatherman_util`. So try adding `leatherman_locale`; you'll also need to manually add any dependencies like boost locale.

dvhwgumby commented:
Indeed, manipulating the various includes are what I do now for Leatherman and hocon-cpp.

However CMake doesn't want users mucking with those files, especially since (who knows?) perhaps some build systems don't use them (all the ones **I** use do, but still). 

target libraries & module names are supposed to manage both includes and library files (.a, .dylib, .so, whatever).  They want us to let CMake do the heavy lifting.  For a long time I thought the poorl named `target_link_libraries` was for `-l` and `-L` but really it is "these are the third party libraries depended on -- make sure cpp can find the headers and ld can find the library objects"

Thus the preferred, and simpler mode is the user doing the merely following:

```
foo.cc:
  #include <boost/locale>
  ....

*CMakeLists.txt:*
  find_package(Boost REQUIRED COMPONENTS locale)

  add_executable(foo foo.cc)
  target_link_libraries (foo Boost::locale)
```

And if I used Leatherman locale:

```
foo.cc:
  #include <leatherman/locale>
  ....

CMakeLists.txt:

  find_package(Leatherman REQUIRED COMPONENTS locale)

  add_executable(foo foo.cc)
  target_link_libraries (foo Leatherman::locale)
```

And let Leatherman manage its own need for Boost etc transparently to the user.  

This is what I wanted to do in a pull request for HOCON but I couldn't get Leatherman to DTRT.  There's enough going on in the Leatherman .cmake files that I'm afraid to break them.

Eventually I will want to compile with VS, which I know nothing about, so I'm eager to be able to tell CMake everything I can at as high a level as I can.

MikaelSmith commented:
That seems like a pattern that may have been undocumented when we wrote the Leatherman CMake files (circa CMake 3.2). So I'm not familiar with it, and don't know off the top of my head why Leatherman wouldn't support it. It's probable Leatherman needs some work to let CMake manage inherited dependencies on targets.

dvhwgumby commented:
Yeah, Cmake's documentation, while surprisingly comprehensive, still manages to be horrible.

I don't mean to encourage gratuitous meta work on your company's CMake infrastructure.  I'm grateful you guys ship module support in lib/cmake/leatherman at all.

If you guys ever do get around to updating the leatherman CMake module definition I'll do that for hocon and send a PR.


> On Aug 15, 2019, at 13:55, Michael Smith <notifications@github.com> wrote:
> 
> That seems like a pattern that may have been undocumented when we wrote the Leatherman CMake files (circa CMake 3.2). So I'm not familiar with it, and don't know off the top of my head why Leatherman wouldn't support it. It's probable Leatherman needs some work to let CMake manage inherited dependencies on targets.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/puppetlabs/leatherman/pull/1?email_source=notifications&email_token=AADOANBEEB4TVBNIIOW67PDQEW7C5A5CNFSM4A4SLQF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4M7EKA#issuecomment-521794088>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AADOANEG6DUZGAE7M4DBD7DQEW7C5ANCNFSM4A4SLQFQ>.
>
